### PR TITLE
feat(registry): add scaleway_registry_namespace_beta resource

### PR DESCRIPTION
**Pull request by VoyTechnology**
Adds container registry and basic documentation associated with it.

### CODE REVIEW

File: scaleway/provider.go [hunk 1]
```diff
@@ -212,6 +212,7 @@ func Provider() terraform.ResourceProvider {
 			"scaleway_security_group_rule":      resourceScalewaySecurityGroupRule(),
 			"scaleway_volume":                   resourceScalewayVolume(),
 			"scaleway_volume_attachment":        resourceScalewayVolumeAttachment(),
+			"scaleway_container_registry":       resourceScalewayContainerRegistry(),
```
Comment (QuentinBrosse) [inline scaleway/provider.go]: We will name it `scaleway_registry_namespace_beta`.

We want to respect this pattern for all new resources: `scaleway_{productName}_{resourceName}[_beta]`. For every new resource, we prefer to release it with `_beta` suffix and remove it when we will be confident about the maturity of the resource.

Note that we have to rename:
- All functions: CRUD, helpers, tests, etc
- Documentation
- Files

Sorry for the inconvenience. 😕

File: scaleway/resource_container_registry.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+	if err != nil {
+		if is404Error(err) {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	_ = d.Set("name", ns.Name)
```
Comment (QuentinBrosse) [inline scaleway/resource_container_registry.go]: nit: you can remove `_ = `, just for consistency.

File: scaleway/resource_container_registry.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	description := d.Get("description").(string)
+	isPublic := d.Get("is_public").(bool)
+
+	if d.HasChange("description") || d.HasChange("is_public") {
+		if _, err := api.UpdateNamespace(&registry.UpdateNamespaceRequest{
+			Region:      region,
+			NamespaceID: id,
+			Description: &description,
```
Comment (QuentinBrosse) [inline scaleway/resource_container_registry.go]:
```go
Description: scw.StringPtr(d.Get("description").(string)),
IsPublic: scw.BoolPtr(d.Get("is_public").(bool)),
```

File: scaleway/resource_container_registry.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+			return nil
+		}
+		return err
+	}
+
+	_ = d.Set("name", ns.Name)
+	_ = d.Set("description", ns.Description)
+	_ = d.Set("organization_id", ns.OrganizationID)
+	_ = d.Set("is_public", ns.IsPublic)
+	_ = d.Set("endpoint", ns.Endpoint)
```
Comment (QuentinBrosse) [inline scaleway/resource_container_registry.go]: Add `d.Set("region", ns.Region)`

File: website/scaleway.erb [hunk 5]
```diff
@@ -124,6 +124,15 @@
           </ul>
         </li>
 
+        <li>
```
Comment (QuentinBrosse) [inline website/scaleway.erb]: - Move it between `Baremetal Resources (Beta)` and `Kubernetes Resources (Beta)`.

File: scaleway/resource_container_registry_test.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+			return err
+		}
+
+		_, err = api.DeleteNamespace(&registry.DeleteNamespaceRequest{
+			NamespaceID: id,
+			Region:      region,
+		})
+
+		if err == nil {
+			return fmt.Errorf("Namespace (%s) still exists", rs.Primary.ID)
```
Comment (QuentinBrosse) [inline scaleway/resource_container_registry_test.go]: nit: `namespace`

File: scaleway/resource_container_registry_test.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+			},
+			{
+				Config: `
+					resource scaleway_container_registry cr01 {
+						name = "test-cr"
+						description = "test container repository"
+						is_public = true
+					}
+				`,
+				Check: resource.ComposeTestCheckFunc(
```
Comment (QuentinBrosse) [inline scaleway/resource_container_registry_test.go]: Add checks for `description` and `is_public`.

File: website/docs/r/container_registry.markdown [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+Creates and manages Scaleway Container Registry. For more information see [the documentation](https://developers.scaleway.com/en/products/registry/api/).
+
+## Examples
+
+### Basic
+
+```hcl
+resource "scaleway_container_registry" "main" {
+    region = "fr-par"
```
Comment (QuentinBrosse) [inline website/docs/r/container_registry.markdown]: Remove `region` in basic example.

File: website/docs/r/container_registry.markdown [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    name = "main_cr"
+    description = "Main container registry"
+    is_public = false
+}
+```
+
+## Arguments Reference
```
Comment (QuentinBrosse) [inline website/docs/r/container_registry.markdown]: Can you move `region` and `organization_id` arguments after `is_public`? The majority of users will use the default ones.

File: website/docs/r/container_registry.markdown [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+~> **Important** Updates to `name` will recreate the namespace.
+
+- `region` - (Defaults to [provider](../index.html#region) `region`). The [region](../guides/regions_and_zones.html#regions) in which the container registry namespace should be created.
+
+- `organization_id` - (Defaults to [provider](../index.html#organization_id) `organization_id`) The ID of the organization the registry is associated with.
+
+- `description` (Optional) The description of the container registry namespace.
+
+- `is_public` (Defaults to `false`) Whether or not the registry should be visible to the public.
```
Comment (AmFlint) [inline website/docs/r/container_registry.markdown]: Whether or not the Registry Images stored in the namespace will be downloadable publicly (docker pull).

File: scaleway/resource_container_registry.go (hunk 4 above)
Comment (loicbourgois) [reply scaleway/resource_container_registry.go]: What about other fields documented at https://developers.scaleway.com/en/products/registry/api/#get-3cb8cf ?

Comment (loicbourgois) [reply scaleway/resource_container_registry.go]: If other read only fields are not Set, why set `endpoint` ?

Comment (QuentinBrosse) [reply scaleway/resource_container_registry.go]: We will expose only `name`, `description`, `organization_id`, `is_public`, `endpoint`, `region` and `id` for the moment. If a user has a use-case who requires another attribute we will talk about it at this moment.

**Comment by QuentinBrosse** [comment]
Hi @VoyTechnology and happy new year!

There is no activity in your PR since its creation on Nov 20, 2019. Are you available to finish it?
If you are not, we will close this PR et reopen another one to continue the work.

Thanks

File: scaleway/helper_container_registry.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
+package scaleway
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/scaleway/scaleway-sdk-go/api/registry/v1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+func containerRegistryWithRegion(d *schema.ResourceData, m interface{}) (*registry.API, scw.Region, error) {
```
[the remark this answers is not in the document]
Comment (VoyTechnology) [reply scaleway/helper_container_registry.go]: Please see #336

File: scaleway/helper_container_registry.go [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+	meta := m.(*Meta)
+	api := registry.NewAPI(meta.scwClient)
+
+	region, err := getRegion(d, meta)
+	return api, region, err
+}
+
+func containerRegistryWithRegionAndID(m interface{}, id string) (*registry.API, scw.Region, string, error) {
```
[the remark this answers is not in the document]
Comment (VoyTechnology) [reply scaleway/helper_container_registry.go]: Please see #336

File: scaleway/provider.go (hunk 1 above)
Comment (VoyTechnology) [reply scaleway/provider.go]: I hope I have covered all of the cases. If there are more please comment on them and I will change it.

**Comment by VoyTechnology** [comment]
I have renamed the resource as requested and rebased the code. Some of the comments are still valid, I am not sure how to continue on about them so I would appreciate extra details.

File: scaleway/helper_container_registry.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
+package scaleway
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/scaleway/scaleway-sdk-go/api/registry/v1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+)
+
+func registryNamespaceWithRegion(d *schema.ResourceData, m interface{}) (*registry.API, scw.Region, error) {
```
Comment (QuentinBrosse) [inline scaleway/helper_container_registry.go]: Rename the function (for consistency) and add a godoc:
```go
// containerRegistryAPIWithRegion returns a new container registry API and the region for a Create request`
func containerRegistryAPIWithRegion(d *schema.ResourceData, m interface{}) (*registry.API, scw.Region, error) {
```

File: scaleway/helper_container_registry.go [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+	meta := m.(*Meta)
+	api := registry.NewAPI(meta.scwClient)
+
+	region, err := getRegion(d, meta)
+	return api, region, err
+}
+
+func registryNamespaceWithRegionAndID(m interface{}, id string) (*registry.API, scw.Region, string, error) {
```
Comment (QuentinBrosse) [inline scaleway/helper_container_registry.go]: Same here (rename + godoc): `registryNamespaceAPIWithRegionAndID`

File: scaleway/resource_registry_namespace_beta.go [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+	}
+}
+
+func resourceScalewayRegistryNamespaceBetaCreate(d *schema.ResourceData, m interface{}) error {
+	api, region, err := registryNamespaceWithRegion(d, m)
+	if err != nil {
+		return err
+	}
+
+	name, ok := d.GetOk("name")
```
Comment (QuentinBrosse) [inline scaleway/resource_registry_namespace_beta.go]: As the name is a unique identifier, we don't want to generate it and let the user explicitly specify it.

> Note that a namespace name is unique in a region. Thus, if another client already has created "test", you can't have it as a namespace.

You have marked the `name` attribute as required, it's fine, now you also have to remove the:
```go
if !ok {
	name = getRandomName("cr")
}
```

File: scaleway/resource_registry_namespace_beta.go [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+	if err != nil {
+		if is404Error(err) {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	_ = d.Set("name", ns.Name)
```
Comment (QuentinBrosse) [inline scaleway/resource_registry_namespace_beta.go]: nit: for consistency, can you can remove `_ =` in all the following lines?

File: scaleway/resource_registry_namespace_beta_test.go [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+		CheckDestroy: testAccCheckScalewayRegistryNamespaceBetaDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: `
+					resource scaleway_registry_namespace_beta cr01 {
+						name = "test-cr"
+					}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayRegistryNamespaceBetaExists("scaleway_registry_namespace_beta.cr01"),
```
Comment (QuentinBrosse) [inline scaleway/resource_registry_namespace_beta_test.go]: Remove empty line

File: scaleway/resource_registry_namespace_beta_test.go [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+			{
+				Config: `
+					resource scaleway_registry_namespace_beta cr01 {
+						name = "test-cr"
+						description = "test container repository"
+						is_public = true
+					}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayRegistryNamespaceBetaExists("scaleway_registry_namespace_beta.cr01"),
```
Comment (QuentinBrosse) [inline scaleway/resource_registry_namespace_beta_test.go]: Remove empty line

File: scaleway/resource_registry_namespace_beta_test.go [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+					testCheckResourceAttrUUID("scaleway_registry_namespace_beta.cr01", "id"),
```
Comment (QuentinBrosse) [inline scaleway/resource_registry_namespace_beta_test.go]: Add checks for `description` and `is_public` attributes.

File: website/docs/r/registry_namespace_beta.markdown [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+
+The following arguments are supported:
+
+- `name` - (Required) The unique name of the container registry namespace.
+
+~> **Important** Updates to `name` will recreate the namespace.
+
+- `description` (Optional) The description of the container registry namespace.
+
+- `is_public` (Defaults to `false`) Whether or not the Registry Images stored in the namespace should be downloadable publicly (docker pull).
```
Comment (QuentinBrosse) [inline website/docs/r/registry_namespace_beta.markdown]: nit: "registry images" (in lowercase)

File: scaleway/resource_registry_namespace_beta_test.go (hunk 17 above)
Comment (VoyTechnology) [reply scaleway/resource_registry_namespace_beta_test.go]: To which empty line are you referring to?

File: scaleway/resource_registry_namespace_beta_test.go (hunk 18 above)
Comment (VoyTechnology) [reply scaleway/resource_registry_namespace_beta_test.go]: I'm sorry but I am not sure what line you are talking about.

File: website/scaleway.erb [hunk 21]
```diff
@@ -109,6 +109,15 @@
...
+          <a href="#">Container Registry Resources</a>
+          <ul class="nav nav-visible">
+            <li>
+              <a href="/docs/providers/scaleway/r/registry_namespace.html">scaleway_registry_namespace_beta</a>
```
Comment (jerome-quere) [inline website/scaleway.erb]: Should be `/docs/providers/scaleway/r/registry_namespace_beta.html`

File: scaleway/resource_registry_namespace_beta_test.go (hunk 17 above)
Comment (QuentinBrosse) [reply scaleway/resource_registry_namespace_beta_test.go]: Sorry, it was a diff issue.

File: scaleway/resource_registry_namespace_beta_test.go (hunk 18 above)
Comment (QuentinBrosse) [reply scaleway/resource_registry_namespace_beta_test.go]: Sorry, it was a diff issue.

**Comment by VoyTechnology** [comment]
The CI is failing because of
```
build github.com/terraform-providers/terraform-provider-scaleway: cannot load github.com/scaleway/scaleway-sdk-go/api/registry/v1: open /home/travis/gopath/src/github.com/terraform-providers/terraform-provider-scaleway/vendor/github.com/scaleway/scaleway-sdk-go/api/registry/v1: no such file or directory
```

I am not sure how does it relate to my PR, since the  `registry/v1` package is available in the version `1.0.0-beta5` used, and I have not changed anything in how the project is set up.

**Comment by QuentinBrosse** [comment]
@VoyTechnology 
- To fix tests, can you try to get the last head version of the `scaleway-sdk-go` with `go get github.com/scaleway/scaleway-sdk-go@master` and/or rebase your branch to `master`?
- Can you also fix [this comment](https://github.com/terraform-providers/terraform-provider-scaleway/pull/330#discussion_r370512650), please?

Once this is done, your PR will be ready to merge. :)

**Comment by VoyTechnology** [comment]
@QuentinBrosse The CI is now passing. Can you PTAL?

**Comment by VoyTechnology** [comment]
Would you like me to squash all the commits before the merge?

**Comment by QuentinBrosse** [comment]
Nop, I will do squash and merge with Github.
Thanks again for your great work @VoyTechnology! 🎉

**Pull request merged**